### PR TITLE
Make union types pretty again

### PR DESCRIPTION
**Pull request by jmorrell**
A first stab at implementing #28 

This looks better than the current formatting to my eye, but I'd be happy for any feedback so I can iterate and improve this 😃

### CODE REVIEW

File: tests/disjoint-union-perf/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -84,13 +84,14 @@ export type TypedNode =
...
  */
 
-export type InferredType = \"unknown\" |
-  \"gender\" |
-  \"enum\" |
-  \"number-or-string\" |
-  \"number\" |
-  \"string\" |
-  \"error\";
+export type InferredType =
```
Comment (jmorrell) [inline tests/disjoint-union-perf/__snapshots__/jsfmt.spec.js.snap]: There is a trailing space after the `=`, and it wasn't clear to me how I could get rid of it...

File: tests/type-destructors/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -69,6 +69,7 @@ var x1: $NonMaybeType<number | string> = true;
 // err, boolean ~> number|string
 var x2: $PropertyType<{ p: number } | { p: string }, \"p\"> = 0;
 // ok, number ~> number|string
-var x3: $PropertyType<{ p: number } |
-  { p: string }, \"p\"> = true; // err, boolean ~> number|string"
+var x3: $PropertyType<
```
Comment (jmorrell) [inline tests/type-destructors/__snapshots__/jsfmt.spec.js.snap]: This is kind of funky... Looks like the trailing comment is throwing it off, but it brings up a point. Do we want the same line-breaking behavior when a union type is used as a type parameter?

```js
var x: TypeWithParameters<{ p: 'long-test-string' } | { p: 'other-long-test-string' } | { p: 'one-more' }>; 
```

File: tests/union-intersection/test.js [hunk 3]
```diff
@@ -13,3 +13,38 @@ function f3(x: T3): T3 { return x; }
 
 type T4 = (A | C) & (B | C);
 function f4(x: T4): T4 { return x; }
+
+type T5 = "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" | "10" | "11" | "12" | "13";
```
Comment (jmorrell) [inline tests/union-intersection/test.js]: NB: I added some more examples to the test

**Comment by jlongster** [comment]
The snapshot changes look good to me. I'm curious what @vjeux thinks, do you like?

**Comment by vjeux** [comment]
I like it, this is what we have on Molten fyi

**Comment by jlongster** [comment]
Awesome, thanks for the work @jmorrell. Thanks for figuring out how `conditionalGroup` works and all that.

File: tests/disjoint-union-perf/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (jlongster) [inline tests/disjoint-union-perf/__snapshots__/jsfmt.spec.js.snap]: Hm, I'm not sure. I think I'm going to revisit declarations again soon, so we'll figure it out.

File: tests/type-destructors/__snapshots__/jsfmt.spec.js.snap (hunk 2 above)
Comment (jlongster) [inline tests/type-destructors/__snapshots__/jsfmt.spec.js.snap]: I'd say it's fine for now. The common case this is solving is worth it and we can figure out these special cases later.

**Pull request merged**